src/rust/src/declarative_asn1/decode.rs

```diff
@@ -57,6 +59,45 @@ fn decode_printable_string<'a>(
     Ok(pyo3::Bound::new(py, PrintableString { inner })?)
 }
 
+fn decode_utc_time<'a>(
+    py: pyo3::Python<'a>,
+    parser: &mut Parser<'a>,
+) -> ParseResult<pyo3::Bound<'a, UtcTime>> {
+    let value = parser.read_element::<asn1::UtcTime>()?;
+    let dt = value.as_datetime();
+
+    let inner = crate::x509::datetime_to_py_utc(py, dt)?
+        .downcast_into::<pyo3::types::PyDateTime>()
+        .map_err(|_| {
+            pyo3::exceptions::PyValueError::new_err(
+                "error converting to datetime object".to_string(),
+            )
+        })?
+        .unbind();
+
+    Ok(pyo3::Bound::new(py, UtcTime { inner })?)
+}
+
+fn decode_generalized_time<'a>(
+    py: pyo3::Python<'a>,
+    parser: &mut Parser<'a>,
+) -> ParseResult<pyo3::Bound<'a, GeneralizedTime>> {
+    let value = parser.read_element::<asn1::GeneralizedTime>()?;
+    let dt = value.as_datetime();
+    let microseconds = value.nanoseconds().map_or(0, |nanos| nanos / 1_000);
```

Comment:
hmm, this is going to truncate off some precision, which means things won't roundtrip. is this a problem?

Reply:
mm yeah that's a good point. I think we should do what `rust-asn1` does, and fail if the decoded fraction exceeds our maximum precision (in our case, 999.999 microseconds).